File: web/src/server/storage.ts
import "server-only";
import { config } from "./config";

/** Supabase Storage REST helpers using the service role key. Ported from
 * backend/app/storage.py. */

function headers(): Record<string, string> {
  return {
    Authorization: `Bearer ${config.supabaseServiceRoleKey}`,
    apikey: config.supabaseServiceRoleKey,
    "Content-Type": "application/json",
  };
}

export interface SignedUpload {
  url: string;
  path: string;
}

export interface StorageObject {
  name: string;
  metadata?: { size?: number };
}

/** POST /storage/v1/object/upload/sign/{bucket}/{path} -> {url, token} */
export async function createSignedUploadUrl(
  bucket: string,
  path: string,
  expiresIn = 900,
): Promise<SignedUpload> {
  const resp = await fetch(
    `${config.supabaseUrl}/storage/v1/object/upload/sign/${bucket}/${path}`,
    { method: "POST", headers: headers(), body: JSON.stringify({ expiresIn }) },
  );
  if (!resp.ok) throw new Error(`createSignedUploadUrl failed: ${resp.status}`);
  const data = await resp.json();
  const signedUrl = data.url ?? data.signedURL ?? data.signedUrl;
  return { url: `${config.supabaseUrl}/storage/v1${signedUrl}`, path };
}

/** POST /storage/v1/object/sign/{bucket}/{path} -> signed GET url */
export async function createSignedDownloadUrl(
  bucket: string,
  path: string,
  expiresIn = 3600,
): Promise<string> {
  const resp = await fetch(`${config.supabaseUrl}/storage/v1/object/sign/${bucket}/${path}`, {
    method: "POST",
    headers: headers(),
    body: JSON.stringify({ expiresIn }),
  });
  if (!resp.ok) throw new Error(`createSignedDownloadUrl failed: ${resp.status}`);
  const data = await resp.json();
  const signedUrl = data.signedURL ?? data.signedUrl;
  return `${config.supabaseUrl}/storage/v1${signedUrl}`;
}

/** POST /storage/v1/object/list/{bucket} -> objects directly under prefix. */
export async function listObjects(
  bucket: string,
  prefix: string,
  limit = 10000,
): Promise<StorageObject[]> {
  const resp = await fetch(`${config.supabaseUrl}/storage/v1/object/list/${bucket}`, {
    method: "POST",
    headers: headers(),
    body: JSON.stringify({ prefix, limit }),
  });
  if (!resp.ok) throw new Error(`listObjects failed: ${resp.status}`);
  return resp.json();
}

/** GET /storage/v1/object/{bucket}/{path} -> ArrayBuffer. */
export async function download(bucket: string, path: string): Promise<ArrayBuffer> {
  const resp = await fetch(`${config.supabaseUrl}/storage/v1/object/${bucket}/${path}`, {
    headers: headers(),
  });
  if (!resp.ok) throw new Error(`download failed: ${resp.status}`);
  return resp.arrayBuffer();
}

/** POST /storage/v1/object/{bucket}/{path} with the file's bytes (upsert). */
export async function upload(
  bucket: string,
  path: string,
  data: Uint8Array | ArrayBuffer,
  contentType = "application/octet-stream",
): Promise<string> {
  const resp = await fetch(`${config.supabaseUrl}/storage/v1/object/${bucket}/${path}`, {
    method: "POST",
    headers: { ...headers(), "Content-Type": contentType, "x-upsert": "true" },
    body: data as BodyInit,
  });
  if (!resp.ok) throw new Error(`upload failed: ${resp.status}`);
  return path;
}

export async function deleteObject(bucket: string, path: string): Promise<void> {
  await fetch(`${config.supabaseUrl}/storage/v1/object/${bucket}/${path}`, {
    method: "DELETE",
    headers: headers(),
  });
}

/** List and delete all objects under a prefix (used for account deletion). */
export async function deletePrefix(bucket: string, prefix: string): Promise<void> {
  const resp = await fetch(`${config.supabaseUrl}/storage/v1/object/list/${bucket}`, {
    method: "POST",
    headers: headers(),
    body: JSON.stringify({ prefix }),
  });
  if (resp.status !== 200) return;
  const objects: StorageObject[] = await resp.json();
  const paths = objects.map((o) => `${prefix}/${o.name}`);
  if (paths.length > 0) {
    await fetch(`${config.supabaseUrl}/storage/v1/object/${bucket}`, {
      method: "DELETE",
      headers: headers(),
      body: JSON.stringify({ prefixes: paths }),
    });
  }
}

File: backend/app/storage.py
"""Supabase Storage REST helpers using the service role key."""

import httpx

from app.config import settings


def _headers() -> dict:
    return {
        "Authorization": f"Bearer {settings.supabase_service_role_key}",
        "apikey": settings.supabase_service_role_key,
        "Content-Type": "application/json",
    }


async def create_signed_upload_url(bucket: str, path: str, expires_in: int = 900) -> dict:
    """POST /storage/v1/object/upload/sign/{bucket}/{path} -> {url, token}."""
    async with httpx.AsyncClient() as client:
        resp = await client.post(
            f"{settings.supabase_url}/storage/v1/object/upload/sign/{bucket}/{path}",
            headers=_headers(),
            json={"expiresIn": expires_in},
            timeout=15.0,
        )
        resp.raise_for_status()
        data = resp.json()
    signed_url = data.get("url") or data.get("signedURL") or data.get("signedUrl")
    return {"url": f"{settings.supabase_url}/storage/v1{signed_url}", "path": path}


async def create_signed_download_url(bucket: str, path: str, expires_in: int = 3600) -> str:
    """POST /storage/v1/object/sign/{bucket}/{path} -> signed GET url."""
    async with httpx.AsyncClient() as client:
        resp = await client.post(
            f"{settings.supabase_url}/storage/v1/object/sign/{bucket}/{path}",
            headers=_headers(),
            json={"expiresIn": expires_in},
            timeout=15.0,
        )
        resp.raise_for_status()
        data = resp.json()
    signed_url = data.get("signedURL") or data.get("signedUrl")
    return f"{settings.supabase_url}/storage/v1{signed_url}"


async def list_objects(bucket: str, prefix: str, limit: int = 10000) -> list[dict]:
    """POST /storage/v1/object/list/{bucket} -> objects directly under prefix.
    Each item includes at least `name` and `metadata.size` for real uploads.
    """
    async with httpx.AsyncClient() as client:
        resp = await client.post(
            f"{settings.supabase_url}/storage/v1/object/list/{bucket}",
            headers=_headers(),
            json={"prefix": prefix, "limit": limit},
            timeout=15.0,
        )
        resp.raise_for_status()
        return resp.json()


async def download(bucket: str, path: str, dest: str) -> str:
    """GET /storage/v1/object/{bucket}/{path} -> writes the object to dest."""
    async with httpx.AsyncClient() as client:
        resp = await client.get(
            f"{settings.supabase_url}/storage/v1/object/{bucket}/{path}",
            headers=_headers(),
            timeout=120.0,
        )
        resp.raise_for_status()
    with open(dest, "wb") as f:
        f.write(resp.content)
    return dest


async def upload(bucket: str, path: str, src: str, content_type: str = "application/octet-stream") -> str:
    """POST /storage/v1/object/{bucket}/{path} with the file's bytes (upsert)."""
    with open(src, "rb") as f:
        data = f.read()
    headers = {**_headers(), "Content-Type": content_type, "x-upsert": "true"}
    async with httpx.AsyncClient() as client:
        resp = await client.post(
            f"{settings.supabase_url}/storage/v1/object/{bucket}/{path}",
            headers=headers,
            content=data,
            timeout=120.0,
        )
        resp.raise_for_status()
    return path


async def delete_object(bucket: str, path: str) -> None:
    async with httpx.AsyncClient() as client:
        await client.request(
            "DELETE",
            f"{settings.supabase_url}/storage/v1/object/{bucket}/{path}",
            headers=_headers(),
            timeout=15.0,
        )


async def delete_prefix(bucket: str, prefix: str) -> None:
    """List and delete all objects under a prefix (used for account deletion)."""
    async with httpx.AsyncClient() as client:
        resp = await client.post(
            f"{settings.supabase_url}/storage/v1/object/list/{bucket}",
            headers=_headers(),
            json={"prefix": prefix},
            timeout=15.0,
        )
        if resp.status_code != 200:
            return
        objects = resp.json()
        paths = [f"{prefix}/{o['name']}" for o in objects]
        if paths:
            await client.request(
                "DELETE",
                f"{settings.supabase_url}/storage/v1/object/{bucket}",
                headers=_headers(),
                json={"prefixes": paths},
                timeout=15.0,
            )
